### PR TITLE
Added typescript code for running bot

### PR DESCRIPTION
**Pull request by YusupovAI**
There is a typescript code for running liquidation bot. The CI is broken now, I am not sure why, will be fixed soon.
Features added:
1. Interactions with `The graph` via graphclient. The .graphclient folder is generated automatically and can be ignored while reviewing.
2. The code for filtering positions to liquidate and building transactions necessary for liquidations.

### CODE REVIEW

File: scripts/index.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+const { AlchemyProvider } = providers;
+import { readFileSync } from "fs";
+import { getAmountsForLiquidty } from "./liquidity-math";
+import { sqrt } from "./math-utils";
+import { liquidate } from "./bot";
+
+dotenv.config({ path: __dirname + "/.env" });
+
+const main = async () => {
+    const provider = new AlchemyProvider("goerli", process.env.ALCHEMY_KEY);
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: Let's use generic JSON RPC provider

File: scripts/index.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    let cdp = new ethers.Contract(
+        process.env.CDP as string,
+        readFileSync("./scripts/abis/Vault.json", "utf-8"),
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: minor: maybe human readable ABI from ethers is a bit better and cleaner practice https://docs.ethers.org/v5/api/utils/abi/formats/#abi-formats--human-readable-abi

File: scripts/index.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                tokens[token1] = await getTokenPriceX96(token1, provider, cdp);
+            }
+        }
+    }
+    return tokens;
+};
+
+const getTokenPriceX96 = async (token: string, provider: ethers.providers.Provider, cdp: ethers.Contract) => {
+    const positionOracle = new ethers.Contract(
+        await cdp.oracle(),
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: I think this call should be cached

File: scripts/index.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        readFileSync("./scripts/abis/UniV3Oracle.json", "utf-8"),
+        provider
+    );
+    const chainlinkOracle = new ethers.Contract(
+        await positionOracle.oracle(),
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: I think this call should be cached

File: scripts/index.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+const liquidationNeeded = async (vault: Vault, tokenPricesX96: any, globalStabilisationFeePerUSDD: BigNumber) => {
+    const capital = getAdjustedCapital(vault, tokenPricesX96);
+    const debt = getOverallDebt(vault, globalStabilisationFeePerUSDD);
+    return capital.lt(debt);
+};
+
+const getAdjustedCapital = (vault: Vault, tokenPricesX96: any) => {
+    // NOTE: here the value of adjusted capital can be a bit lower than actual value
+    // because position accumulated fees are not considered
+    const DENOMINATOR = BigNumber.from(10).pow(9);
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: Let's make `BigNumber.from(2).pow(96)` and `BigNumber.from(10).pow(9)` top level constants

File: scripts/query.graphql [hunk 6]
```diff
@@ -0,0 +1,20 @@
+query Vaults {
+  vaults {
+    id
+    vaultDebt
```
Comment (k1rill-fedoseev) [inline scripts/query.graphql:4]: `WHERE vaultDebt > 0`?

File: scripts/bot.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    console.log(
+        "Trying to liquidate vault %s with repayment %s and debt %s",
+        vault.id,
+        toRepay.toString(),
+        actualDebt.toString()
+    );
+    const nfts = await cdp.vaultNftsById(vault.id);
+    const { swapAddresses, swapData } = await buildAllSwapData(
+        bot.address,
+        await cdp.positionManager(),
```
Comment (k1rill-fedoseev) [inline scripts/bot.ts]: I think this call should be cached

File: scripts/index.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    );
+    const vaults: Vault[] = (await execute(VaultsDocument, {})).data.vaults;
+    const tokenPricesX96 = await getNeededPricesX96(vaults, provider, cdp);
+    const globalStabilisationFeePerUSDD = await cdp.globalStabilisationFeePerUSDD();
+    for (let i = 0; i < vaults.length; ++i) {
+        const vault = vaults[i];
+        if (await liquidationNeeded(vault, tokenPricesX96, globalStabilisationFeePerUSDD)) {
+            await liquidate(vault, cdp, provider);
+        }
+    }
```
Comment (k1rill-fedoseev) [inline scripts/index.ts]: Let's run this in a loop, once each N seconds?

File: scripts/bot.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        nfts,
+        provider
+    );
+    const tx = await bot.connect(signer).liquidate(
```
Comment (k1rill-fedoseev) [inline scripts/bot.ts]: Let's add some extra % to estimated gasLimit and gasPrice

Comment (k1rill-fedoseev) [inline scripts/bot.ts]: Let's use NonceManager from ethers for auto-managing sender nonces

File: scripts/bot.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            debt: toRepay,
+            nfts,
+            swapAddresses,
+            swapData,
+            positionManager: process.env.POSITION_MANAGER,
+            cdp: cdp.address,
+        },
+        signer.address
+    );
+    await tx.wait();
```
Comment (k1rill-fedoseev) [inline scripts/bot.ts]: I think it's better if this would be omitted

File: scripts/index.ts (hunk 8 above)
Comment (YusupovAI) [reply scripts/index.ts]: I think the loop can be done in some deployment config. For example in cron config

Comment (k1rill-fedoseev) [reply scripts/index.ts]: I don't think it's a good idea to run this using cron. I will create an extra overhead, reduce flexibility and make unnecessary RPC requests. So let's make this loop inside the process.

File: scripts/bot.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    const { liquidationPremiumD } = await cdp.protocolParams();
+    const toRepay = overallCollateral.mul(DENOMINATOR.sub(liquidationPremiumD)).div(DENOMINATOR);
+    console.log(
+        "Trying to liquidate vault %s with repayment %s and debt %s",
+        vault.id,
+        toRepay.toString(),
+        actualDebt.toString()
+    );
+    const nfts = await cdp.vaultNftsById(vault.id);
+    const positionManager = await cdp.positionManager();
```
Comment (k1rill-fedoseev) [inline scripts/bot.ts]: Is it the same as `process.env.POSITION_MANAGER`

File: scripts/index.ts (hunk 2 above)
Comment (YusupovAI) [reply scripts/index.ts]: Is there any tool to generate human readable ABI?

Comment (k1rill-fedoseev) [reply scripts/index.ts]: Don't know, I would write all the methods I need by hand, seems like we need only a few of them. Files are also fine, you can keep them if you want.

**Review by k1rill-fedoseev — APPROVED** [review]

**Pull request merged**